### PR TITLE
feat: Add state queries for balance and nonce retrieval of any known block height

### PR DESCRIPTION
**Pull request by RomanHodulak**
### Description
Implements historical lookup of balances and nonces from the blockchain state.

### Changes
- Add state queries for balance and nonce retrieval of any known block height.

### Testing
- Unit test of endpoint
- Unit test of state actor
- Unit test of state queries

### CODE REVIEW

File: moved/src/move_execution/state.rs
Comment (birchmd) [inline moved/src/move_execution/state.rs:65]: The amount of work here is linear in the height of the request, which will be totally unacceptable after more than a few thousand blocks. I think if our strategy is to store change sets then we also need to store occasional complete state checkpoints to avoid starting from 0 on each query.

I still think the better approach (for now) is to not store change sets at all and instead use the merkle store for history. The idea is that we are already storing the complete set of merkle trie nodes which are part of the merkle tries of all past blocks. To look up a state element at height H we look up the block at H, get its state root R and then index into the merkle trie with root R to find the leaf node we need. This make the amount of work independent of block height. The trade off is that queries will be sensitive to the depth of the trie, but this should grow much slower than H so that it will be fine in the medium term. And avoiding direct merkle queries something we can address in the future with a more clever flat storage implementation.

**Review by birchmd — CHANGES_REQUESTED** [review]

File: engine-api/src/jsonrpc.rs
Comment (bulenttastan) [inline engine-api/src/jsonrpc.rs:36]: nit: Capitalization of `block` to `Block`

File: engine-api/src/methods/get_balance.rs
Comment (bulenttastan) [inline engine-api/src/methods/get_balance.rs:21]: Nice 😎

File: moved/src/state_actor/queries.rs
Comment (bulenttastan) [inline moved/src/state_actor/queries.rs:116]: nit: it should be probably called `IntoVersionedKV` (V is capital) or the long way `IntoVersionedKeyValue`.

File: moved/src/state_actor/queries.rs
Comment (bulenttastan) [inline moved/src/state_actor/queries.rs:212]: Should it be `crate::Result<Nonce>` or `Option<Nonce>`?

File: moved/src/state_actor/queries.rs
Comment (bulenttastan) [inline moved/src/state_actor/queries.rs:93]: Nice, using tag as the block height version 🙌

**Review by bulenttastan — COMMENTED** [review]
Overall it looks very good. I'll let @birchmd to take a look before approving.

**Review by birchmd — COMMENTED** [review]
Looks good overall! I left a few additional comments. Rebase onto `main` is also needed to resolve conflicts.

File: moved/src/state_actor/queries.rs
Comment (RomanHodulak) [reply moved/src/state_actor/queries.rs:204]: I think that is already happening. It only fails on invariant violation (eth_token_invariant_violation).

I have a test for it called `test_query_fetches_zero_balance_for_non_existent_account`

File: engine-api/src/methods/get_balance.rs
Comment (bulenttastan) [inline engine-api/src/methods/get_balance.rs:53]: We should just rename the `inner_execute` input to `height: BlockNumberOrTag`. Same applies to get_nonce.

File: engine-api/src/methods/get_nonce.rs
Comment (bulenttastan) [inline engine-api/src/methods/get_nonce.rs:110]: We could simply use `AccountAddress::ONE`

File: engine-api/src/methods/mod.rs
Comment (bulenttastan) [inline engine-api/src/methods/mod.rs:159]: Feels like `address` should be the first param, then comes the `height`.

File: moved/src/state_actor/queries.rs
Comment (RomanHodulak) [reply moved/src/state_actor/queries.rs:643]: There is another test for it, this one checks older balance.

Comment (RomanHodulak) [reply moved/src/state_actor/queries.rs:643]: Oh now I see it, you're right.

**Review by bulenttastan — APPROVED** [review]
Looks good. This was a big PR 😮‍💨

**Review by birchmd — APPROVED** [review]
Approving to unblock. Historical lookups are an important feature! I still think the new `OnTx` and `OnTxBatch` might be unnecessary complexity, but we can always iterate on the design later.

**Pull request merged**
